src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java

```diff
@@ -0,0 +1,221 @@
+            switch (event.getKind()) {
+                case OVERFLOW: acceptOverflow(); break;
+                case CREATED: acceptCreated(event.calculateFullPath()); break;
+                case DELETED: acceptDeleted(event.calculateFullPath()); break;
+                case MODIFIED: break;
+            }
+        }
+
+        private void acceptOverflow() {
+            for (var childWatch : childWatches.values()) {
+                reportOverflowTo(childWatch);
+            }
+        }
+
+        private void acceptCreated(Path fullPath) {
+            if (Files.isDirectory(fullPath)) {
+                var childWatch = openChildWatch(fullPath);
+                // Events in the newly created directory might have been missed
+                // between its creation and setting up its watch. So, generate
+                // an `OVERFLOW` event for the watch.
+                reportOverflowTo(childWatch);
+            }
+        }
+
+        private void acceptDeleted(Path fullPath) {
+            try {
+                closeChildWatch(fullPath);
+            } catch (IOException e) {
+                logger.error("Could not close (nested) file tree watch for: {} ({})", fullPath, e);
+            }
+        }
+
+        private void reportOverflowTo(JDKFileTreeWatch childWatch) {
+            var overflow = new WatchEvent(WatchEvent.Kind.OVERFLOW,
+                childWatch.rootPath, childWatch.relativePathParent);
+            childWatch.handleEvent(overflow);
+        }
+    }
+
+    private JDKFileTreeWatch openChildWatch(Path child) {
+        Function<Path, JDKFileTreeWatch> newChildWatch = p -> new JDKFileTreeWatch(
+            rootPath, rootPath.relativize(child), exec, eventHandler);
+
+        var childWatch = childWatches.computeIfAbsent(child, newChildWatch);
+        try {
+            childWatch.startIfFirstTime();
+        } catch (IOException e) {
+            logger.error("Could not open (nested) file tree watch for: {} ({})", child, e);
+        }
+        return childWatch;
+    }
+
+    private void closeChildWatch(Path child) throws IOException {
+        var childWatch = childWatches.remove(child);
+        if (childWatch != null) {
+            childWatch.close();
+        }
+    }
+
+    // -- JDKBaseWatch --
+
+    @Override
+    public WatchScope getScope() {
+        return WatchScope.PATH_AND_ALL_DESCENDANTS;
+    }
+
+    @Override
+    public void handleEvent(WatchEvent event) {
+        internal.handleEvent(event);
+    }
+
+    @Override
+    public synchronized void close() throws IOException {
+        IOException firstFail = null;
+        var children = childWatches.keySet().iterator();
+        while (true) {
+            try {
+                // First, close all child watches
+                if (children.hasNext()) {
+                    closeChildWatch(children.next());
+                }
+                // Last, close the internal watch
+                else {
+                    internal.close();
```

Comment:
I think we should fist close our internal watch, to make sure we don't get new childeren registered on different threads, and then clear out chunks of the `children`, remove them from the map, and then loop untill map is empty. Maybe even set a `volatile boolean` field to mark this watchers as closed.